Cyclic symlink workaround

daschuer opened the pull request:
This is a PR, that features the behaviour discussed in 
https://github.com/mixxxdj/mixxx/pull/708

Tracks, previously discovered in duplicated directories are no treated as deleted. 
Scanning still rejects directories that are already track by their canonical path. 
In 1.11, duplicates where silently removed if one of the duplicated track is not reachable during the scan. This has changed in this PR, now a missing duplicated Track is listed as missing tracks as well.

uklotzde commented:
@daschuer Unfortunately I experience spurious crashes during scanning at different points when starting Mixxx with a fresh/empty home directory. The crashes do not occur with the original PR code.

Here are 2 examples:

Thread 2 (Thread 0x7ffec6bae700 (LWP 12503)):
#0  0x0000000000000000 in  ()
#1  0x00007ffff55dc3ac in QApplicationPrivate::notify_helper(QObject*, QEvent*) () at /lib64/libQtGui.so.4
#2  0x00007ffff55e3206 in QApplication::notify(QObject*, QEvent*) ()
    at /lib64/libQtGui.so.4
#3  0x0000000000afac02 in MixxxApplication::notify(QObject*, QEvent*) (this=0x7fffffffdd30, target=0x7ffec002a460, event=0x7ffeb800fa60)
    at src/mixxxapplication.cpp:138
#4  0x00007ffff4bebf1d in QCoreApplication::notifyInternal(QObject*, QEvent*) () at /lib64/libQtCore.so.4
#5  0x00007ffff4bef536 in QCoreApplicationPrivate::sendPostedEvents(QObject*, int, QThreadData*) () at /lib64/libQtCore.so.4
#6  0x00007ffff4c1c35e in postEventSourceDispatch(_GSource*, int (*)(void*), void*) () at /lib64/libQtCore.so.4
#7  0x00007fffedeaca8a in g_main_context_dispatch () at /lib64/libglib-2.0.so.0
---Type <return> to continue, or q <return> to quit---
#8  0x00007fffedeace20 in g_main_context_iterate.isra ()
    at /lib64/libglib-2.0.so.0
#9  0x00007fffedeacecc in g_main_context_iteration ()
    at /lib64/libglib-2.0.so.0
#10 0x00007ffff4c1c4ce in QEventDispatcherGlib::processEvents(QFlags<QEventLoop::ProcessEventsFlag>) () at /lib64/libQtCore.so.4
#11 0x00007ffff4bea791 in QEventLoop::processEvents(QFlags<QEventLoop::ProcessEventsFlag>) () at /lib64/libQtCore.so.4
#12 0x00007ffff4beab05 in QEventLoop::exec(QFlags<QEventLoop::ProcessEventsFlag>) () at /lib64/libQtCore.so.4
#13 0x00007ffff4ada289 in QThread::exec() () at /lib64/libQtCore.so.4
#14 0x0000000000a9a4c3 in LibraryScanner::run() (this=0x45431f0)
    at src/library/scanner/libraryscanner.cpp:159

Thread 3 (Thread 0x7ffec63ad700 (LWP 12056)):
#0  0x0000000000bcbea2 in QList<QString>::contains(QString const&) const (this=0x7ffec63ac800, t=...) at /usr/include/QtCore/qlist.h:872
#1  0x0000000000bc7301 in SoundSourceProxy::initialize(QString const&) (qFilename=...) at src/soundsourceproxy.cpp:190
#2  0x0000000000bc7a8e in SoundSourceProxy::SoundSourceProxy(QString, QSharedPointer<SandboxSecurityToken>) (this=0x7ffec63aca30, qFilename=..., pToken=...)
    at src/soundsourceproxy.cpp:75
#3  0x0000000000bf591e in TrackInfoObject::parse(bool) (this=this@entry=0x7ffeb8---Type <return> to continue, or q <return> to quit---
009140, parseCoverArt=parseCoverArt@entry=true) at src/trackinfoobject.cpp:171
#4  0x0000000000bf6da9 in TrackInfoObject::initialize(bool, bool) (this=this@entry=0x7ffeb8009140, parseHeader=parseHeader@entry=true, parseCoverArt=parseCoverArt@entry=true) at src/trackinfoobject.cpp:135
#5  0x0000000000bf7120 in TrackInfoObject::TrackInfoObject(QString const&, QSharedPointer<SandboxSecurityToken>, bool, bool) (this=0x7ffeb8009140, file=..., pToken=..., parseHeader=<optimized out>, parseCoverArt=<optimized out>)
    at src/trackinfoobject.cpp:53
#6  0x0000000000a908b6 in ImportFilesTask::run() (this=0x7ffeb8012d60)
    at src/library/scanner/importfilestask.cpp:50

daschuer commented:
@uklotzde, Thank you for testing. 
I have found two race-conditions, fixed in the latest commit. They are only bad in specific timing conditions. I am able to reproduce these crashes on my low end device, however, my back-traces are looking quite different. The effected code is also part of 1.12 branch and seams to be the cause of 
https://bugs.launchpad.net/mixxx/+bug/1443451

Luckily the issue was more easy to reproduce in the new second scanning path. 

Please test again. I hope that your crashes are also fixed now.

daschuer commented:
I have found some more issues. Hopfully everything fine now.
